### PR TITLE
[dynamo][guards-cpp-refactor] TENSOR_ALIASING guard

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120098
* #120096
* #120093
* #120123
* #120119
* #120091
* #120089
* #120068
* #120067
* #120065
* __->__ #120064
* #120062
* #120061
* #120060
* #119833
* #119827
* #119822

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by anijain2305** [comment]
@jansel Ready for another review.

File: torch/csrc/dynamo/guards.cpp [hunk 1]
```diff
@@ -976,6 +976,34 @@ class RelationalGuard : public LeafGuard {
...
+      : RelationalGuard(verbose_code_parts), _is_first_call(true) {}
+
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    if (_is_first_call) {
+      _first_tensor = value;
+      _is_first_call = false;
+      return true;
+    }
+    bool result = _first_tensor == value;
+    reset_state();
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:994]: We should be able to avoid a virtual method call here for slightly faster performance.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/csrc/dynamo/guards.cpp [hunk 2]
```diff
@@ -976,6 +976,34 @@ class RelationalGuard : public LeafGuard {
...
+    return result;
+  }
+
+  void reset_state() override {
```
Comment (jansel) [inline torch/csrc/dynamo/guards.cpp:998]: Adding final here will remove virtual method call I believe

Comment (anijain2305) [reply torch/csrc/dynamo/guards.cpp:1005]: you are right! Adding final skips the v-table indirect call - https://stackoverflow.com/questions/28332918/virtual-function-efficiency-and-the-final-keyword

**Comment by anijain2305** [comment]
@jansel Ready for another review.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**